Add OpenTherm component (part 3: rest of the sensors)

olegtarasov opened the pull request:
# What does this implement/fix?

This is the third (and final) part of OpenTherm component. It contains the code to support all the sensors, inputs and outputs. Sensors configured in YAML are transformed to a set of C++ macros, which in turn generate the final code that handles sensors.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- Part 1: #6645 
- Part 2: #7529 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#3800

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# An extremely minimal configuration which only enables you to set the boiler's
# water temperature setpoint as a number.

esphome:
  name: thermostat-number-minimal

esp8266:
  board: d1_mini

opentherm:
  in_pin: IN_PIN
  out_pin: OUT_PIN

number:
  - platform: opentherm
    t_set:
      name: "Boiler Control setpoint"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

FreeBear-nc committed:
Add boiler messages from the most recent OT function matrix
https://www.opentherm.eu/wp-content/uploads/2016/06/OpenTherm-Function-Matrix-v1_0.xls

FreeBear-nc committed:
Remove duplicate message ID:100

FreeBear-nc committed:
Add HVAC messages from the most recent OT function matrix
https://www.opentherm.eu/wp-content/uploads/2016/06/OpenTherm-Function-Matrix-v1_0.xls

FreeBear-nc committed:
Add solar messages from the most recent OT function matrix
https://www.opentherm.eu/wp-content/uploads/2016/06/OpenTherm-Function-Matrix-v1_0.xls

FreeBear-nc committed:
silly, silly little error - comma omitted.

olegtarasov committed:
Final part of OpenTherm sensors

olegtarasov committed:
Added back tests for all platforms

olegtarasov commented:
Hi @jesserockz! The final part of OpenTherm integration is ready for review. Thanks!

FreeBear-nc committed:
Merge branch 'esphome:dev' into opentherm-solar

olegtarasov committed:
Merge branch 'dev' into opentherm-sensors2

FreeBear-nc committed:
Add new sensor: fan_speed_setpoint - This reads ID35:hb
Modify sensor: fan_speed - This now reads ID35:lb

Both return a value as revs per second. Multiply by 60 to get RPM

FreeBear-nc committed:
Merge branch 'opentherm-sensors2' into opentherm-fan

FreeBear-nc committed:
Whitespace changes to keep clang-format happy.

olegtarasov committed:
Merge pull request #2 from FreeBear-nc/opentherm-fan

Opentherm fan

olegtarasov committed:
Merge branch 'dev' into opentherm-sensors2

FreeBear-nc committed:
Extend the parse_u8_* functions to automatically multiply value by 60.
This is only used by ID35{lb,hb} and saves having to use a multiply
filter in the sensor.yaml

FreeBear-nc committed:
Default icon missing for fan speed. As pointed out by Simone Rossetto
it should be mdi:fan.

olegtarasov committed:
Merge pull request #3 from FreeBear-nc/opentherm-fan

Opentherm fan

olegtarasov committed:
Merge branch 'dev' into opentherm-sensors2

jesserockz reviewed (commented):
How does the hub know if a switch (or other control type entity) has been changed by the user? I do not see the `add_on_state_callback`s being added.

I find it very difficult to follow and understand this component due to the heavy use of macros.

FreeBear-nc committed:
Allow message_data to be overriden by a yaml entry.
This allows users to return data in a different format. e.g.:

  fan_speed:
  name: "Boiler fan speed"
  message_data: "u16" # overrides the default u8_lb_60 message.

olegtarasov commented:
@jesserockz The hub doesn't need to know the exact moment when control entity changes because all sensors, switches and outputs are sent to OpenTherm device constantly in a loop. It's required by OpenTherm standard to communicate with device every 1 second even when there is no change in controls. So the approach when we send all the messages configured in yaml in a loop adds to component's robustness (a message can fail due to protocol error, and we don't need to use complicated retry logic — the same message will be sent again eventually).

> I find it very difficult to follow and understand this component due to the heavy use of macros.

I agree, but I see no other way to add class members dynamically, and adding > 100 class members for all the available sensors at once seems very wasteful. What's more, most of them likely won't be supported by end user's boiler.

olegtarasov committed:
Merge branch 'dev' into opentherm-sensors2

olegtarasov committed:
Sync with develop

olegtarasov committed:
Merge branch 'free-bear' into opentherm-sensors2

olegtarasov committed:
Merge branch 'opentherm-sensors2' into opentherm-dev

olegtarasov committed:
Added an ability to specify master OpenTherm version (message id 124)

olegtarasov committed:
Fixed linter warnings

olegtarasov committed:
Merge branch 'dev' into opentherm-sensors2

jesserockz commented on tests/components/opentherm/common.yaml:
```diff
@@ -74,3 +105,86 @@ sensor:
+    ch2_active:
+      name: "Boiler Central Heating 2 active"
+
+climate:
+  - platform: pid
+    name: "Central heating"
+    heat_output: t_set
+    default_target_temperature: 20
+    sensor: ch_room_temperature
+    control_parameters:
+      kp: 0.7
+      ki: 0.003
+  - platform: pid
+    name: "Central heating (Circuit 2)"
+    heat_output: t_set_ch2
+    default_target_temperature: 18
+    sensor: ch2_room_temperature
+    control_parameters:
+      kp: 0.4
+      ki: 0.004
```
Adding this here wont achieve anything and if the `pid` component has a breaking change, these files wont be run. They should be contained to only itself and dependencies
```suggestion
```

jesserockz commented on tests/components/opentherm/common.yaml:
```diff
@@ -74,3 +105,86 @@ sensor:
       name: "OTC heat curve ratio upper bound"
     otc_hc_ratio_lb:
       name: "OTC heat curve ratio lower bound"
+
+  - platform: homeassistant
+    id: ch_room_temperature
+    entity_id: sensor.temperature
+    filters:
+      # Push room temperature every second to update PID parameters
+      - heartbeat: 1s
+  - platform: homeassistant
+    id: ch2_room_temperature
+    entity_id: sensor.other_temperature
+    filters:
+      # Push room temperature every second to update PID parameters
+      - heartbeat: 1s
```
Same with these
```suggestion
```

jesserockz commented on tests/components/opentherm/common.yaml:
```diff
@@ -12,10 +12,41 @@ opentherm:
   cooling_enable: false
   otc_active: false
   ch2_active: true
+  t_room: ch_room_temperature
```
Just use one of the opentherm sensors here as this is a simple compile test and doesnt actually run

jesserockz commented on esphome/components/opentherm/switch/switch.cpp:
```diff
@@ -0,0 +1,27 @@
+}
+
+void OpenthermSwitch::dump_config() {
+  esphome::switch_::log_switch("opentherm.switch", "", "OpenTherm Switch", this);
```
```suggestion
  LOG_SWITCH(TAG, "", "OpenTherm Switch", this);
```

jesserockz commented on esphome/components/opentherm/switch/switch.cpp:
```diff
@@ -0,0 +1,27 @@
+
+namespace esphome {
+namespace opentherm {
+
```
```suggestion

static const char *const TAG = "opentherm.switch";

```

jesserockz commented on esphome/components/opentherm/switch/switch.cpp:
```diff
@@ -0,0 +1,27 @@
+  auto restored = this->get_initial_state_with_restore_mode();
+  bool state = false;
+  if (!restored.has_value()) {
+    ESP_LOGD("opentherm.switch", "Couldn't restore state for OpenTherm switch '%s'", this->get_name().c_str());
```
Please update to add TAG to all the cpp files and use that for logging
```suggestion
    ESP_LOGD(TAG, "Couldn't restore state for OpenTherm switch '%s'", this->get_name().c_str());
```

jesserockz commented on esphome/components/opentherm/switch/switch.cpp:
```diff
@@ -0,0 +1,27 @@
+  if (!restored.has_value()) {
+    ESP_LOGD("opentherm.switch", "Couldn't restore state for OpenTherm switch '%s'", this->get_name().c_str());
+  } else {
+    ESP_LOGD("opentherm.switch", "Restored state for OpenTherm switch '%s': %d", this->get_name().c_str(),
```
```suggestion
    ESP_LOGD(TAG, "Restored state for OpenTherm switch '%s': %d", this->get_name().c_str(),
```

olegtarasov committed:
Merge branch 'dev' into opentherm-sensors2

olegtarasov committed:
Apply suggestions from code review

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

olegtarasov committed:
More fixes from code review

olegtarasov committed:
Removed a comment

olegtarasov committed:
Fixed OpenTherm test

jesserockz committed:
Use LOG_NUMBER
